### PR TITLE
Adds FUNC_E_PLATFORM to override host OS and architecture used to run Envoy

### LINKED ISSUE #299: Error: couldn't find version "1.18.3" for platform "darwin/arm64"
**Describe the bug**
I do not have an Envoy configuration file, so I am using the option to start the admin port. When I run `func-e run` using the JSON provided in the documentation, I get an error.

**To Reproduce**
1. Run ```curl https://func-e.io/install.sh | bash -s -- -b /usr/local/bin``` to install func-e
2. Run ```func-e run --config-yaml "admin: {address: {socket_address: {address: '127.0.0.1', port_value: 9901}}}"``` to use the admin port

**Expected behavior**
I expected to get the latest version of Envoy. MacOS should be able to run application for amd64 on Rosetta 2.

**Screenshots**

**Environment (please complete the relevant information):**
 - OS: MacOS 11.3.1 (M1)
 - Envoy Version: n/a

### PR DESCRIPTION
**Pull request by karl-cardenas-coding**
This PR adds a CLI flag for specifying a target platform for the Envoy install. This also includes the ability to specify a target platform through an environment variable.

Closes #299

### CODE REVIEW

**Comment by karl-cardenas-coding** [comment]
@codefromthecrypt Can you take a peek and let me know if I close to what you are thinking? I will add test cases once we have the code hashed out.

File: internal/envoy/install.go [hunk 1]
```diff
@@ -35,39 +35,39 @@ import (
...
+			return "", fmt.Errorf("couldn't find releaseDate of version %q for platform %q: %w", v, o.Platform, err)
 		}
 		if err = os.MkdirAll(installPath, 0750); err != nil {
 			return "", fmt.Errorf("unable to create directory %q: %w", installPath, err)
 		}
 
-		fmt.Fprintln(o.Out, "downloading", tarballURL)                                   //nolint
-		if err = untarEnvoy(ctx, installPath, tarballURL, sha256Sum, p, v); err != nil { //nolint
+		fmt.Fprintln(o.Out, "downloading", tarballURL)                                                              //nolint
+		if err = untarEnvoy(ctx, installPath, tarballURL, sha256Sum, version.Platform(o.Platform), v); err != nil { //nolint
```
Comment (codefromthecrypt) [inline internal/envoy/install.go:70]: if you make o.Platform literally `version.Platform` then there is less doing this.

File: internal/cmd/use.go [hunk 2]
```diff
@@ -42,7 +42,7 @@ $ func-e use %s`, envoy.CurrentVersionWorkingDirFile, envoy.CurrentVersionHomeDi
 		Before: validateVersionArg,
 		Action: func(c *cli.Context) error {
 			v := version.Version(c.Args().First())
-			if _, err := envoy.InstallIfNeeded(c.Context, o, globals.CurrentPlatform, v); err != nil {
```
Comment (codefromthecrypt) [inline internal/cmd/use.go:45]: deleting `globals.CurrentPlatform` will help show all the call sites to change

File: internal/cmd/app.go [hunk 3]
```diff
@@ -62,7 +62,15 @@ func NewApp(o *globals.GlobalOpts) *cli.App {
 			DefaultText: globals.DefaultEnvoyVersionsURL,
 			Destination: &envoyVersionsURL,
 			EnvVars:     []string{"ENVOY_VERSIONS_URL"},
-		}}
+		},
+		&cli.StringFlag{
+			Name:        "platform",
+			Usage:       "target platform to install Envoy",
+			DefaultText: globals.DefaultPlatform,
+			Destination: &envoyVersionsURL,
```
Comment (codefromthecrypt) [inline internal/cmd/app.go:70]:
```suggestion
			Destination: &envoyPlatform,
```

then later you would check if "" and set o.Platform accordingly

**Review by codefromthecrypt — COMMENTED** [review]
I think you are on the right track!

File: internal/cmd/app.go (hunk 3 above)
Comment (karl-cardenas-coding) [reply internal/cmd/app.go]: @codefromthecrypt  The  TestInstallIfNeeded is getting the incorrect date from `string(test.FakeReleaseDate)`
```
internal/envoy - (platform-flag) $ go test ./...
--- FAIL: TestInstallIfNeeded (0.09s)
    install_test.go:168:
        	Error Trace:	install_test.go:168
        	Error:      	Not equal:
        	            	expected: "2020-12-30"
        	            	actual  : "2020-12-31"

        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1 +1 @@
        	            	-2020-12-30
        	            	+2020-12-31
        	Test:       	TestInstallIfNeeded
FAIL
FAIL	github.com/tetratelabs/func-e/internal/envoy	2.264s
```
The test snippet failing.
```go
	f, err := os.Stat(versionDir)
	require.NoError(t, err)
	require.Equal(t, f.ModTime().Format("2006-01-02"), string(test.FakeReleaseDate))
```
 
Should I change `FakeReleaseDate = version.ReleaseDate("2020-12-31")` ?

Comment (codefromthecrypt) [reply internal/cmd/app.go]: hmm maybe insert `.UTC()` between ModTime and Format?

Comment (codefromthecrypt) [reply internal/cmd/app.go]: if this doesn't work, I'd suggest to focus on the other tests and push whatever you can. I can take a look and help on this one!

Comment (karl-cardenas-coding) [reply internal/cmd/app.go]: @codefromthecrypt Yeah, I'm seeing various failed test due to time offset by 1. I'm wrapping up the test for the platform arg and fixing the Usage.MD test.  I'll have a push here in a bit that I'll tag you into review/

Comment (karl-cardenas-coding) [reply internal/cmd/app.go]: @codefromthecrypt Okay, I think I'm all done on my end. If you have a moment, can you check out the few checks that are failing due to the time offset please ?

Comment (codefromthecrypt) [reply internal/cmd/app.go]: Yeap sure will and try to help you land this so https://github.com/tetratelabs/func-e/pull/301 doesn't drift it!

**Comment by codefromthecrypt** [comment]
ok I added a polishing commit to save some back and forth. Mainly, we don't want to advertise using func-e routinely in a way to override platform as it can lead to mismatch which would hurt support. So, I moved the doc into an "advanced" section.

I think the timezone thing is unrelated to this change, so not addressing it here.

Thanks tons!

**Comment by codefromthecrypt** [comment]
nevermind travis it is accidentally building both the PR and your branch. This will fix it https://github.com/tetratelabs/func-e/pull/302

**Comment by codefromthecrypt** [comment]
ok fingers crossed we're green now

**Review by mathetake — APPROVED** [review]
Looks good to me

**Comment by codefromthecrypt** [comment]
cool. I will release this right away and also update our Homebrew recipe :D

**Comment by mathetake** [comment]
Thanks @karl-cardenas-coding for your contribution:)

**Comment by karl-cardenas-coding** [comment]
@mathetake Happy to help. Thanks for merging this so quickly.

**Pull request merged**